<!doctype html><html lang=en class=no-js> <head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><meta name=description content=AGAT><meta name=author content="Jacques Dainat"><link href=https://github.com/NBISweden/AGAT/agat_how_does_it_work/ rel=canonical><link href=../agat_for_you/ rel=prev><link href=../how_to_cite/ rel=next><link rel=icon href=../assets/images/favicon.png><meta name=generator content="mkdocs-1.6.1, mkdocs-material-9.6.16"><title>How does AGAT work? - AGAT</title><link rel=stylesheet href=../assets/stylesheets/main.7e37652d.min.css><link rel=stylesheet href=../assets/stylesheets/palette.06af60db.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script></head> <body dir=ltr data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#how-does-agat-work class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <header class=md-header data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=.. title=AGAT class="md-header__button md-logo" aria-label=AGAT data-md-component=logo> <img src=../img/NBIS.png alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> AGAT </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> How does AGAT work? </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to dark mode" type=radio name=__palette id=__palette_0> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg> </label> <input class=md-option data-md-color-media data-md-color-scheme=slate data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to light mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_0 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12s-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg> </label> </form> <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg> </label> <nav class=md-search__options aria-label=Search> <a href=javascript:void(0) class="md-search__icon md-icon" title=Share aria-label=Share data-clipboard data-clipboard-text data-md-component=search-share tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7s-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91s2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08"/></svg> </a> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg> </button> </nav> <div class=md-search__suggest data-md-component=search-suggest></div> </form> <div class=md-search__output> <div class=md-search__scrollwrap tabindex=0 data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list role=presentation></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/NBISweden/AGAT title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 7.0.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path fill=currentColor d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg> </div> <div class=md-source__repository> AGAT </div> </a> </div> </nav> </header> <div class=md-container data-md-component=container> <nav class=md-tabs aria-label=Tabs data-md-component=tabs> <div class=md-grid> <ul class=md-tabs__list> <li class="md-tabs__item md-tabs__item--active"> <a href=.. class=md-tabs__link> AGAT Home </a> </li> <li class=md-tabs__item> <a href=../gxf/ class=md-tabs__link> The GTF/GFF formats </a> </li> <li class=md-tabs__item> <a href=../topological-sorting-of-gff-features/ class=md-tabs__link> AGAT vs other tools </a> </li> <li class=md-tabs__item> <a href=../tools/agat_convert_sp_gxf2gxf/ class=md-tabs__link> AGAT Tools </a> </li> <li class=md-tabs__item> <a href=../howto/how_to_extract_sequences/ class=md-tabs__link> How to </a> </li> <li class=md-tabs__item> <a href=../troubleshooting/ class=md-tabs__link> Troubleshooting </a> </li> </ul> </div> </nav> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary md-nav--lifted md-nav--integrated" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=.. title=AGAT class="md-nav__button md-logo" aria-label=AGAT data-md-component=logo> <img src=../img/NBIS.png alt=logo> </a> AGAT </label> <div class=md-nav__source> <a href=https://github.com/NBISweden/AGAT title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 7.0.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path fill=currentColor d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg> </div> <div class=md-source__repository> AGAT </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_1 checked> <div class="md-nav__link md-nav__container"> <a href=.. class="md-nav__link "> <span class=md-ellipsis> AGAT Home </span> </a> <label class="md-nav__link " for=__nav_1 id=__nav_1_label tabindex> <span class="md-nav__icon md-icon"></span> </label> </div> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_1_label aria-expanded=true> <label class=md-nav__title for=__nav_1> <span class="md-nav__icon md-icon"></span> AGAT Home </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../install/ class=md-nav__link> <span class=md-ellipsis> Installation </span> </a> </li> <li class=md-nav__item> <a href=../agat_for_you/ class=md-nav__link> <span class=md-ellipsis> What can AGAT do for you? </span> </a> </li> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> <span class=md-ellipsis> How does AGAT work? </span> <span class="md-nav__icon md-icon"></span> </label> <a href=./ class="md-nav__link md-nav__link--active"> <span class=md-ellipsis> How does AGAT work? </span> </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#the-agat-parser-used-by-_sp_-prefix-tools-standardisation-to-create-gxf-files-compliant-to-any-tool class=md-nav__link> <span class=md-ellipsis> The AGAT parser / used by _sp_ prefix tools / Standardisation to create GXF files compliant to any tool </span> </a> <nav class=md-nav aria-label="The AGAT parser / used by _sp_ prefix tools / Standardisation to create GXF files compliant to any tool"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#what-performs-the-agat-parser class=md-nav__link> <span class=md-ellipsis> What performs the AGAT parser </span> </a> </li> <li class=md-nav__item> <a href=#the-data-structure class=md-nav__link> <span class=md-ellipsis> The data structure </span> </a> </li> <li class=md-nav__item> <a href=#how-does-the-agat-parser-work class=md-nav__link> <span class=md-ellipsis> How does the AGAT parser work </span> </a> </li> <li class=md-nav__item> <a href=#particular-case class=md-nav__link> <span class=md-ellipsis> Particular case </span> </a> <nav class=md-nav aria-label="Particular case"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#a-level1-feature-type-missing-and-no-parentgene_id class=md-nav__link> <span class=md-ellipsis> A. Level1 feature type missing and no Parent/gene_id </span> </a> </li> <li class=md-nav__item> <a href=#b-level1-and-level2-feature-types-missing-only-level3-features class=md-nav__link> <span class=md-ellipsis> B. Level1 and Level2 feature types missing (Only Level3 features!) </span> </a> <nav class=md-nav aria-label="B. Level1 and Level2 feature types missing (Only Level3 features!)"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#1-case-with-parentid-transcript_idgene_id-relationships class=md-nav__link> <span class=md-ellipsis> 1. Case with Parent/ID transcript_id/gene_id relationships. </span> </a> </li> <li class=md-nav__item> <a href=#2-case-without-parentid-transcript_idgene_id-relationships-only-common-attribute-approach-to-parse-the-file-can-be-used class=md-nav__link> <span class=md-ellipsis> 2. Case without Parent/ID transcript_id/gene_id relationships. Only common attribute approach to parse the file can be used. </span> </a> </li> <li class=md-nav__item> <a href=#3-in-the-extreme-case-where-you-have-only-one-type-of-feature-you-may-decide-to-use-the-id-as-common-attribute class=md-nav__link> <span class=md-ellipsis> 3. In the extreme case where you have only one type of feature, you may decide to use the ID as common attribute. </span> </a> </li> <li class=md-nav__item> <a href=#4-case-where-you-have-only-one-type-of-feature-and-some-feature-have-parent-attributes-and-some-other-have-common-attributes class=md-nav__link> <span class=md-ellipsis> 4. Case where you have only one type of feature, and some feature have Parent attributes and some other have common attributes. </span> </a> </li> </ul> </nav> </li> </ul> </nav> </li> </ul> </nav> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../how_to_cite/ class=md-nav__link> <span class=md-ellipsis> How to cite? </span> </a> </li> <li class=md-nav__item> <a href=../why_agat/ class=md-nav__link> <span class=md-ellipsis> Why AGAT? </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2> <label class=md-nav__link for=__nav_2 id=__nav_2_label tabindex=0> <span class=md-ellipsis> The GTF/GFF formats </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_2_label aria-expanded=false> <label class=md-nav__title for=__nav_2> <span class="md-nav__icon md-icon"></span> The GTF/GFF formats </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../gxf/ class=md-nav__link> <span class=md-ellipsis> The GTF/GFF formats </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_3> <label class=md-nav__link for=__nav_3 id=__nav_3_label tabindex=0> <span class=md-ellipsis> AGAT vs other tools </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_3_label aria-expanded=false> <label class=md-nav__title for=__nav_3> <span class="md-nav__icon md-icon"></span> AGAT vs other tools </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../topological-sorting-of-gff-features/ class=md-nav__link> <span class=md-ellipsis> Topological sorting of gff features </span> </a> </li> <li class=md-nav__item> <a href=../gff_to_bed/ class=md-nav__link> <span class=md-ellipsis> GFF to BED conversion </span> </a> </li> <li class=md-nav__item> <a href=../gff_to_gtf/ class=md-nav__link> <span class=md-ellipsis> GFF to GTF conversion </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_4> <label class=md-nav__link for=__nav_4 id=__nav_4_label tabindex=0> <span class=md-ellipsis> AGAT Tools </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_4_label aria-expanded=false> <label class=md-nav__title for=__nav_4> <span class="md-nav__icon md-icon"></span> AGAT Tools </label> <ul class=md-nav__list data-md-scrollfix> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_4_1> <label class=md-nav__link for=__nav_4_1 id=__nav_4_1_label tabindex=0> <span class=md-ellipsis> Slurp (sp prefix) </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_4_1_label aria-expanded=false> <label class=md-nav__title for=__nav_4_1> <span class="md-nav__icon md-icon"></span> Slurp (sp prefix) </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../tools/agat_convert_sp_gxf2gxf/ class=md-nav__link> <span class=md-ellipsis> agat_convert_sp_gxf2gxf.pl </span> </a> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_4_1_2> <label class=md-nav__link for=__nav_4_1_2 id=__nav_4_1_2_label tabindex=0> <span class=md-ellipsis> Convert </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=3 aria-labelledby=__nav_4_1_2_label aria-expanded=false> <label class=md-nav__title for=__nav_4_1_2> <span class="md-nav__icon md-icon"></span> Convert </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../tools/agat_convert_bed2gff/ class=md-nav__link> <span class=md-ellipsis> agat_convert_bed2gff.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_convert_embl2gff/ class=md-nav__link> <span class=md-ellipsis> agat_convert_embl2gff.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_convert_genscan2gff/ class=md-nav__link> <span class=md-ellipsis> agat_convert_genscan2gff.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_convert_mfannot2gff/ class=md-nav__link> <span class=md-ellipsis> agat_convert_mfannot2gff.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_convert_minimap2_bam2gff/ class=md-nav__link> <span class=md-ellipsis> agat_convert_minimap2_bam2gff.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_convert_sp_gff2bed/ class=md-nav__link> <span class=md-ellipsis> agat_convert_sp_gff2bed.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_convert_sp_gff2gtf/ class=md-nav__link> <span class=md-ellipsis> agat_convert_sp_gff2gtf.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_convert_sp_gff2tsv/ class=md-nav__link> <span class=md-ellipsis> agat_convert_sp_gff2tsv.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_convert_sp_gff2zff/ class=md-nav__link> <span class=md-ellipsis> agat_convert_sp_gff2zff.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_sp_to_tabulated/ class=md-nav__link> <span class=md-ellipsis> agat_sp_to_tabulated.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_sp_alignment_output_style/ class=md-nav__link> <span class=md-ellipsis> agat_sp_alignment_output_style.pl </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_4_1_3> <label class=md-nav__link for=__nav_4_1_3 id=__nav_4_1_3_label tabindex=0> <span class=md-ellipsis> Add </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=3 aria-labelledby=__nav_4_1_3_label aria-expanded=false> <label class=md-nav__title for=__nav_4_1_3> <span class="md-nav__icon md-icon"></span> Add </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../tools/agat_sp_add_intergenic_regions/ class=md-nav__link> <span class=md-ellipsis> agat_sp_add_intergenic_regions.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_sp_add_introns/ class=md-nav__link> <span class=md-ellipsis> agat_sp_add_introns.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_sp_add_splice_sites/ class=md-nav__link> <span class=md-ellipsis> agat_sp_add_splice_sites.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_sp_add_start_and_stop/ class=md-nav__link> <span class=md-ellipsis> agat_sp_add_start_and_stop.pl.pl </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_4_1_4> <label class=md-nav__link for=__nav_4_1_4 id=__nav_4_1_4_label tabindex=0> <span class=md-ellipsis> Aggregate </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=3 aria-labelledby=__nav_4_1_4_label aria-expanded=false> <label class=md-nav__title for=__nav_4_1_4> <span class="md-nav__icon md-icon"></span> Aggregate </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../tools/agat_sp_complement_annotations/ class=md-nav__link> <span class=md-ellipsis> agat_sp_complement_annotations.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_sp_merge_annotations/ class=md-nav__link> <span class=md-ellipsis> agat_sp_merge_annotations.pl </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_4_1_5> <label class=md-nav__link for=__nav_4_1_5 id=__nav_4_1_5_label tabindex=0> <span class=md-ellipsis> Compare </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=3 aria-labelledby=__nav_4_1_5_label aria-expanded=false> <label class=md-nav__title for=__nav_4_1_5> <span class="md-nav__icon md-icon"></span> Compare </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../tools/agat_sp_compare_two_BUSCOs/ class=md-nav__link> <span class=md-ellipsis> agat_sp_compare_two_BUSCOs.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_sp_compare_two_annotations/ class=md-nav__link> <span class=md-ellipsis> agat_sp_compare_two_annotations.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_sp_sensitivity_specificity/ class=md-nav__link> <span class=md-ellipsis> agat_sp_sensitivity_specificity.pl </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_4_1_6> <label class=md-nav__link for=__nav_4_1_6 id=__nav_4_1_6_label tabindex=0> <span class=md-ellipsis> Extract </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=3 aria-labelledby=__nav_4_1_6_label aria-expanded=false> <label class=md-nav__title for=__nav_4_1_6> <span class="md-nav__icon md-icon"></span> Extract </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../tools/agat_sp_extract_attributes/ class=md-nav__link> <span class=md-ellipsis> agat_sp_extract_attributes.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_sp_extract_sequences/ class=md-nav__link> <span class=md-ellipsis> agat_sp_extract_sequences.pl </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_4_1_7> <label class=md-nav__link for=__nav_4_1_7 id=__nav_4_1_7_label tabindex=0> <span class=md-ellipsis> Filter </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=3 aria-labelledby=__nav_4_1_7_label aria-expanded=false> <label class=md-nav__title for=__nav_4_1_7> <span class="md-nav__icon md-icon"></span> Filter </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../tools/agat_sp_filter_by_ORF_size/ class=md-nav__link> <span class=md-ellipsis> agat_sp_filter_by_ORF_size.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_sp_filter_by_locus_distance/ class=md-nav__link> <span class=md-ellipsis> agat_sp_filter_by_locus_distance.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_sp_filter_by_mrnaBlastValue/ class=md-nav__link> <span class=md-ellipsis> agat_sp_filter_by_mrnaBlastValue.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_sp_filter_feature_by_attribute_presence/ class=md-nav__link> <span class=md-ellipsis> agat_spfilter_feature_by_attribute_presence.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_sp_filter_feature_by_attribute_value/ class=md-nav__link> <span class=md-ellipsis> agat_sp_filter_feature_by_attribute_value.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_sp_filter_feature_from_keep_list/ class=md-nav__link> <span class=md-ellipsis> agat_sp_filter_feature_from_keep_list.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_sp_filter_feature_from_kill_list/ class=md-nav__link> <span class=md-ellipsis> agat_sp_filter_feature_from_kill_list.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_sp_filter_gene_by_intron_numbers/ class=md-nav__link> <span class=md-ellipsis> agat_sp_filter_gene_by_intron_numbers.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_sp_filter_gene_by_length/ class=md-nav__link> <span class=md-ellipsis> agat_sp_filter_gene_by_length.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_sp_filter_incomplete_gene_coding_models/ class=md-nav__link> <span class=md-ellipsis> agat_sp_filter_incomplete_gene_coding_models.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_sp_filter_record_by_coordinates/ class=md-nav__link> <span class=md-ellipsis> agat_sp_filter_record_by_coordinates.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_sp_keep_longest_isoform/ class=md-nav__link> <span class=md-ellipsis> agat_sp_keep_longest_isoform.pl </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_4_1_8> <label class=md-nav__link for=__nav_4_1_8 id=__nav_4_1_8_label tabindex=0> <span class=md-ellipsis> Flag </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=3 aria-labelledby=__nav_4_1_8_label aria-expanded=false> <label class=md-nav__title for=__nav_4_1_8> <span class="md-nav__icon md-icon"></span> Flag </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../tools/agat_sp_flag_premature_stop_codons/ class=md-nav__link> <span class=md-ellipsis> agat_sp_flag_premature_stop_codons.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_sp_flag_short_introns/ class=md-nav__link> <span class=md-ellipsis> agat_sp_flag_short_introns.pl </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_4_1_9> <label class=md-nav__link for=__nav_4_1_9 id=__nav_4_1_9_label tabindex=0> <span class=md-ellipsis> Fix </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=3 aria-labelledby=__nav_4_1_9_label aria-expanded=false> <label class=md-nav__title for=__nav_4_1_9> <span class="md-nav__icon md-icon"></span> Fix </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../tools/agat_sp_fix_cds_phases/ class=md-nav__link> <span class=md-ellipsis> agat_sp_fix_cds_phases.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_sp_fix_features_locations_duplicated/ class=md-nav__link> <span class=md-ellipsis> agat_sp_fix_features_locations_duplicated.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_sp_fix_fusion/ class=md-nav__link> <span class=md-ellipsis> agat_sp_fix_fusion.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_sp_fix_longest_ORF/ class=md-nav__link> <span class=md-ellipsis> agat_sp_fix_longest_ORF.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_sp_fix_overlaping_genes/ class=md-nav__link> <span class=md-ellipsis> agat_sp_fix_overlaping_genes.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_sp_fix_small_exon_from_extremities/ class=md-nav__link> <span class=md-ellipsis> agat_fix_small_exon_from_extremities.pl </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_4_1_10> <label class=md-nav__link for=__nav_4_1_10 id=__nav_4_1_10_label tabindex=0> <span class=md-ellipsis> Manage </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=3 aria-labelledby=__nav_4_1_10_label aria-expanded=false> <label class=md-nav__title for=__nav_4_1_10> <span class="md-nav__icon md-icon"></span> Manage </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../tools/agat_sp_manage_IDs/ class=md-nav__link> <span class=md-ellipsis> agat_sp_manage_IDs.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_sp_manage_UTRs/ class=md-nav__link> <span class=md-ellipsis> agat_sp_manage_UTRs.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_sp_manage_attributes/ class=md-nav__link> <span class=md-ellipsis> agat_sp_manage_attributes.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_sp_manage_functional_annotation/ class=md-nav__link> <span class=md-ellipsis> agat_sp_manage_functional_annotation.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_sp_manage_introns/ class=md-nav__link> <span class=md-ellipsis> agat_sp_manage_introns.pl </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_4_1_11> <label class=md-nav__link for=__nav_4_1_11 id=__nav_4_1_11_label tabindex=0> <span class=md-ellipsis> Statistics </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=3 aria-labelledby=__nav_4_1_11_label aria-expanded=false> <label class=md-nav__title for=__nav_4_1_11> <span class="md-nav__icon md-icon"></span> Statistics </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../tools/agat_sp_statistics/ class=md-nav__link> <span class=md-ellipsis> agat_sp_statistics.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_sp_functional_statistics/ class=md-nav__link> <span class=md-ellipsis> agat_sp_functional_statistics.pl </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_4_1_12> <label class=md-nav__link for=__nav_4_1_12 id=__nav_4_1_12_label tabindex=0> <span class=md-ellipsis> Other </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=3 aria-labelledby=__nav_4_1_12_label aria-expanded=false> <label class=md-nav__title for=__nav_4_1_12> <span class="md-nav__icon md-icon"></span> Other </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../tools/agat_sp_kraken_assess_liftover/ class=md-nav__link> <span class=md-ellipsis> agat_sp_kraken_assess_lift_coverage.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_sp_list_short_introns/ class=md-nav__link> <span class=md-ellipsis> agat_sp_list_short_introns.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_sp_load_function_from_protein_align/ class=md-nav__link> <span class=md-ellipsis> agat_sp_load_function_from_protein_align.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_sp_move_attributes_within_records/ class=md-nav__link> <span class=md-ellipsis> agat_sp_move_attributes_within_records.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_sp_prokka_fix_fragmented_gene_annotations/ class=md-nav__link> <span class=md-ellipsis> agat_sp_prokka_fragmented_gene_annotations.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_sp_separate_by_record_type/ class=md-nav__link> <span class=md-ellipsis> agat_sp_separate_by_record_type.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_sp_split_by_level2_feature/ class=md-nav__link> <span class=md-ellipsis> agat_sp_split_by_level2_feature.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_sp_clipN_seqExtremities_and_fixCoordinates/ class=md-nav__link> <span class=md-ellipsis> agat_sp_clipN_seqExtremities_and_fixCoordinates.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_sp_ensembl_output_style/ class=md-nav__link> <span class=md-ellipsis> agat_sp_ensembl_output_style.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_sp_Prokka_inferNameFromAttributes/ class=md-nav__link> <span class=md-ellipsis> agat_sp_Prokka_inferNameFromAttributes.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_sp_webApollo_compliant/ class=md-nav__link> <span class=md-ellipsis> agat_sp_webApollo_compliant.pl </span> </a> </li> </ul> </nav> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_4_2> <label class=md-nav__link for=__nav_4_2 id=__nav_4_2_label tabindex=0> <span class=md-ellipsis> Sequential (sq prefix) </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_4_2_label aria-expanded=false> <label class=md-nav__title for=__nav_4_2> <span class="md-nav__icon md-icon"></span> Sequential (sq prefix) </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../tools/agat_sq_add_attributes_from_tsv/ class=md-nav__link> <span class=md-ellipsis> agat_sq_add_attributes_from_tsv.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_sq_add_hash_tag/ class=md-nav__link> <span class=md-ellipsis> agat_sq_add_hash_tag.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_sq_add_locus_tag/ class=md-nav__link> <span class=md-ellipsis> agat_sq_add_locus_tag.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_sq_filter_feature_from_fasta/ class=md-nav__link> <span class=md-ellipsis> agat_sq_filter_feature_from_fasta.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_sq_list_attributes/ class=md-nav__link> <span class=md-ellipsis> agat_sq_list_attributes.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_sq_manage_IDs/ class=md-nav__link> <span class=md-ellipsis> agat_sq_manage_ID.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_sq_manage_attributes/ class=md-nav__link> <span class=md-ellipsis> agat_sp_manage_attributes.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_sq_mask/ class=md-nav__link> <span class=md-ellipsis> agat_sq_mask.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_sq_remove_redundant_entries/ class=md-nav__link> <span class=md-ellipsis> agat_sq_remove_redundant_entries.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_sq_repeats_analyzer/ class=md-nav__link> <span class=md-ellipsis> agat_sq_repeats_analyzer.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_sq_reverse_complement/ class=md-nav__link> <span class=md-ellipsis> agat_sq_reverse_complement.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_sq_rfam_analyzer/ class=md-nav__link> <span class=md-ellipsis> agat_sq_rfam_analyzer.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_sq_split/ class=md-nav__link> <span class=md-ellipsis> agat_sq_split.pl </span> </a> </li> <li class=md-nav__item> <a href=../tools/agat_sq_stat_basic/ class=md-nav__link> <span class=md-ellipsis> agat_sq_stat_basic.pl </span> </a> </li> </ul> </nav> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_5> <label class=md-nav__link for=__nav_5 id=__nav_5_label tabindex=0> <span class=md-ellipsis> How to </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_5_label aria-expanded=false> <label class=md-nav__title for=__nav_5> <span class="md-nav__icon md-icon"></span> How to </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../howto/how_to_extract_sequences/ class=md-nav__link> <span class=md-ellipsis> How to extract sequences? </span> </a> </li> <li class=md-nav__item> <a href=../howto/how_to_aggregate_annotation/ class=md-nav__link> <span class=md-ellipsis> How to aggregate several annotations? </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_6> <label class=md-nav__link for=__nav_6 id=__nav_6_label tabindex=0> <span class=md-ellipsis> Troubleshooting </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_6_label aria-expanded=false> <label class=md-nav__title for=__nav_6> <span class="md-nav__icon md-icon"></span> Troubleshooting </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../troubleshooting/ class=md-nav__link> <span class=md-ellipsis> Troubleshooting </span> </a> </li> </ul> </nav> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <a href=https://github.com/NBISweden/AGAT/edit/main/docs/agat_how_does_it_work.md title="Edit this page" class="md-content__button md-icon" rel=edit> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M10 20H6V4h7v5h5v3.1l2-2V8l-6-6H6c-1.1 0-2 .9-2 2v16c0 1.1.9 2 2 2h4zm10.2-7c.1 0 .3.1.4.2l1.3 1.3c.2.2.2.6 0 .8l-1 1-2.1-2.1 1-1c.1-.1.2-.2.4-.2m0 3.9L14.1 23H12v-2.1l6.1-6.1z"/></svg> </a> <a href=https://github.com/NBISweden/AGAT/raw/main/docs/agat_how_does_it_work.md title="View source of this page" class="md-content__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M17 18c.56 0 1 .44 1 1s-.44 1-1 1-1-.44-1-1 .44-1 1-1m0-3c-2.73 0-5.06 1.66-6 4 .94 2.34 3.27 4 6 4s5.06-1.66 6-4c-.94-2.34-3.27-4-6-4m0 6.5a2.5 2.5 0 0 1-2.5-2.5 2.5 2.5 0 0 1 2.5-2.5 2.5 2.5 0 0 1 2.5 2.5 2.5 2.5 0 0 1-2.5 2.5M9.27 20H6V4h7v5h5v4.07c.7.08 1.36.25 2 .49V8l-6-6H6a2 2 0 0 0-2 2v16a2 2 0 0 0 2 2h4.5a8.2 8.2 0 0 1-1.23-2"/></svg> </a> <h1 id=how-does-agat-work>How does AGAT work?<a class=headerlink href=#how-does-agat-work title="Permanent link">#</a></h1> <p>All tools taking GFF/GTF as input can be divided in two groups: <code>_sp_</code> and <code>_sq_</code>.</p> <ul> <li>Tools with <code>_sp_</code> prefix</li> </ul> <p>_sp_ stands for SLURP. Those tools will charge the file in memory in a specific data structure. It has a memory cost but makes life smoother. Indeed, it allows to perform complicated tasks in a more time efficient way ( Any features can be accessed at any time by AGAT). Moreover, it allows to fix all potential errors in the limit of the possibilities given by the format itself. See the AGAT parser section for more information about it. </p> <ul> <li>with <code>_sq_</code> prefix</li> </ul> <p>_sq_ stands for SEQUENTIAL. Those tools will read and process GFF/GTF files from the top to the bottom, line by line, performing tasks on the fly. This is memory efficient but the sanity check of the file is minimum. Those tools are not intended to perform complex tasks.</p> <h2 id=the-agat-parser-used-by-_sp_-prefix-tools-standardisation-to-create-gxf-files-compliant-to-any-tool>The AGAT parser / used by _sp_ prefix tools / Standardisation to create GXF files compliant to any tool<a class=headerlink href=#the-agat-parser-used-by-_sp_-prefix-tools-standardisation-to-create-gxf-files-compliant-to-any-tool title="Permanent link">#</a></h2> <p>The first step of AGAT' tools with the _sp_ prefix of is to fix the file to standardize it. (e.g. a file containing only exon will be modified to create mRNA and gene features). To perform this task AGAT parses and slurps the entire data into a specific data structure. Below you will find more information about peculiarity of this data structure, and the parsing approach used.</p> <h3 id=what-performs-the-agat-parser>What performs the AGAT parser<a class=headerlink href=#what-performs-the-agat-parser title="Permanent link">#</a></h3> <ul> <li>It creates missing parental features. (e.g if a level2 or level3 feature do not have parental feature(s) we create the missing level2 and/or level1 feature(s)). </li> <li>It creates missing mandatory attributes (ID and/or Parent). </li> <li>It fixes identifier to be uniq. </li> <li>It removes duplicated features (same position, same ID, same Parent). </li> <li>It expands level3 features sharing multiple parents (e.g if one exon has list of multiple parent mRNA in its Parent attribute, one exon per parent with uniq ID will be created. </li> <li>It fixes feature location errors (e.g an mRNA spanning over its gene location, we fix the gene location). </li> <li>It adds UTR if possible (CDS and exon present). </li> <li>It adds exon if possible (CDS has to be present). </li> <li>It groups features together (if related features are spread at different places in the file). </li> </ul> <h3 id=the-data-structure>The data structure<a class=headerlink href=#the-data-structure title="Permanent link">#</a></h3> <p>The method create a hash structure containing all the data in memory. We call it OMNISCIENT. The OMNISCIENT structure is a three levels structure: <div class=highlight><pre><span></span><code>$omniscient{level1}{tag_l1}{level1_id} = feature &lt;= tag could be gene, match  
$omniscient{level2}{tag_l2}{idY} = @featureListL2 &lt;= tag could be mRNA,rRNA,tRNA,etc. idY is a level1_id (know as Parent attribute within the level2 feature). The @featureList is a list to be able to manage isoform cases.  
$omniscient{level3}{tag_l3}{idZ} =  @featureListL3 &lt;= tag could be exon,cds,utr3,utr5,etc. idZ is the ID of a level2 feature (know as Parent attribute within the level3 feature). The @featureList is a list to be able to put all the feature of a same tag together.  
</code></pre></div></p> <h3 id=how-does-the-agat-parser-work>How does the AGAT parser work<a class=headerlink href=#how-does-the-agat-parser-work title="Permanent link">#</a></h3> <p>The AGAT parser phylosophy will use several approach to understand the links/relationships betwen the featrures: </p> <ul> <li>1) Parse by Parent/child relationship or gene_id/transcript_id relationship.</li> <li>2) ELSE Parse by a common tag (an attribute value shared by feature that must be grouped together. By default we are using locus_tag but can be set by parameter). </li> <li>3) ELSE Parse sequentially (mean group features in a bucket, and the bucket change at each level2 feature, and bucket are join in a common tag at each new L1 feature). </li> </ul> <p>To resume by priority of way to parse: <strong>Parent/child or gene_id/transcript_id relationship &gt; common attribute/tag &gt; sequential.</strong> </p> <p><img alt src=../img/agat_parsing_overview.jpg width=800px></p> <p>The parser may used only one or a mix of these approaches according of the peculiarity of the gtf/gff file you provide.</p> <p><strong>1. Parsing approach 1: by Parent/child relationship</strong></p> <p>Example of Parent/ID relationship used by the GFF format:</p> <pre><code>chr12   HAVANA  gene    100 500 .   +   .   ID=gene1
chr12   HAVANA  transcript  100 500 .   +   .   ID=transcript1;Parent=gene1
chr12   HAVANA  exon    100 500 .   +   .   ID=exon1;Parent=transcript1  
chr12   HAVANA  CDS 100 500 .   +   0   ID=cds-1;Parent=transcript1
</code></pre> <p>Example of gene_id/transcript_id relationship used by the GTF format:</p> <pre><code>chr12   HAVANA  gene    100 500 .   +   .   gene_id "gene1";
chr12   HAVANA  transcript  100 500 .   +   .   gene_id "gene1"; transcript_id "transcript1";
chr12   HAVANA  exon    100 500 .   +   .   gene_id "gene1"; transcript_id "transcript1"; exon_id=exon1;
chr12   HAVANA  CDS 100 500 .   +   0   gene_id "gene1"; transcript_id "transcript1"; cds_id=cds-1;
</code></pre> <p><strong>2. ELSE Parsing approach 2: by a common attribute/tag</strong> </p> <p>a common attribute (or common tag) is an attribute value shared by feature that must be grouped together. AGAT uses default attributes (<code>gene_id</code> and <code>locus_tag</code>) displayed in the log but can be set by the user modifying the AGAT configuration file <code>agat_config.yaml</code>.<br> You can modify the <code>agat_config.yaml</code> either running <code>agat config --expose</code> to access it (it will be copied in the current directory) and then modifying it manually; or running <code>agat config --expose --locus_tag attribute_name</code> that will copy the <code>agat_config.yaml</code> locally with the modification of the <code>locus_tag</code> parameter accordingly.</p> <p>Example of relationship made using a common tag (here locus_tag):</p> <pre><code>chr12   HAVANA  gene    100 500 .   +   .   locus_tag="gene1"
chr12   HAVANA  transcript  100 500 .   +   .   locus_tag="gene1";ID="transcript1"
chr12   HAVANA  exon    100 500 .   +   .   locus_tag="gene1";ID=exon1;
chr12   HAVANA  CDS 100 500 .   +   0   locus_tag="gene1";ID=cds-1;
</code></pre> <p><strong>3. ELSE Parsing approach 3: sequentially</strong></p> <p>Reading from top to the botom of the file, level3 features (e.g. exon, CDS, UTR) are attached to the last level2 feature (e.g. mRNA) met, and level2 feature are attached to the last L1 feature (e.g. gene) met. To see the list of features of each level see the feature_levels.yaml file (In the share folder in the github repo or using <code>agat levels --expose</code>).</p> <p>Example of relationship made sequentially:</p> <pre><code>chr12   HAVANA  gene    100 500 .   +   .   ID="aaa"
chr12   HAVANA  transcript  100 500 .   +   .   ID="bbb"
chr12   HAVANA  exon    100 500 .   +   .   ID="ccc"
chr12   HAVANA  CDS 100 500 .   +   0   ID="ddd"
chr12   HAVANA  gene    1000    5000    .   +   .   ID="xxx"
chr12   HAVANA  transcript  1000    5000    .   +   .   ID="yyy"
chr12   HAVANA  exon    1000    5000    .   +   .   ID="zzz"
chr12   HAVANA  CDS 1000    5000    .   +   0   ID="www"
</code></pre> <p>/!\ Cases with only level3 features (i.e rast or some prokka files), sequential parsing may not work as expected if Parent/ID gene_id/transcript_id attributes are missing. Indeed all features will be the child of only one newly created Parent. To create a parent per feature or group of features, a common tag must be used to group them correctly (by default gene_id and locus_tag but you can set up the ones of your choice). See <a href=#particular-case>Particular case</a>.</p> <h3 id=particular-case>Particular case<a class=headerlink href=#particular-case title="Permanent link">#</a></h3> <p>Below you will find more information about peculiar GXF files and how the AGAT parser behaves and uses the different parsing approaches.</p> <h4 id=a-level1-feature-type-missing-and-no-parentgene_id>A. Level1 feature type missing and no Parent/gene_id<a class=headerlink href=#a-level1-feature-type-missing-and-no-parentgene_id title="Permanent link">#</a></h4> <p>If you have isoforms (for Eukaryote organism) in your files and the <code>common attribute</code> used is not set properly you can end up with isoforms having independent parent gene features. See below for more details.</p> <p>Here an example of three transcripts from two different genes (isoforms exist - testA.gff):</p> <pre><code>chr12   HAVANA  transcript  100 500 .   +   .   ID="bbb";common_tag="gene1";transcript_id="transcript1";gene_info="gene1"
chr12   HAVANA  exon    100 500 .   +   .   ID="ccc";common_tag="gene1"
chr12   HAVANA  CDS 100 500 .   +   0   ID="ddd";common_tag="gene1"
chr12   HAVANA  transcript  100 600 .   +   .   ID="bbb2";common_tag="gene1";transcript_id="transcript2";gene_info="gene1"
chr12   HAVANA  exon    100 600 .   +   .   ID="ccc2";common_tag="gene1"
chr12   HAVANA  CDS 100 600 .   +   0   ID="ddd2";common_tag="gene1"
chr12   HAVANA  transcript  1000    5000    .   +   .   ID="yyy";common_tag="gene2";transcript_id="transcript3";gene_info="gene2"
chr12   HAVANA  exon    1000    5000    .   +   .   ID="zzz";common_tag="gene2"
chr12   HAVANA  CDS 1000    5000    .   +   0   ID="www";common_tag="gene2"
</code></pre> <ul> <li>/!\ Be careful in Eukaryote annotation containing isoforms. Indeed AGAT will create a gene feature by transcript. As in the example this is wrong because transcript1 and transcript2 should be attached to the same gene: </li> </ul> <p><code>agat_convert_sp_gxf2gxf.pl --gff testA.gff</code></p> <pre><code>chr12   HAVANA  gene    100 500 .   +   .   ID=nbisL1-transcript-1;common_tag="gene1";gene_info="gene1";transcript_id="transcript1"
chr12   HAVANA  transcript  100 500 .   +   .   ID="bbb";Parent=nbisL1-transcript-1;common_tag="gene1";gene_info="gene1";transcript_id="transcript1"
chr12   HAVANA  exon    100 500 .   +   .   ID="ccc";Parent="bbb";common_tag="gene1"
chr12   HAVANA  CDS 100 500 .   +   0   ID="ddd";Parent="bbb";common_tag="gene1"
chr12   HAVANA  gene    100 600 .   +   .   ID=nbisL1-transcript-2;common_tag="gene1";gene_info="gene1";transcript_id="transcript2"
chr12   HAVANA  transcript  100 600 .   +   .   ID="bbb2";Parent=nbisL1-transcript-2;common_tag="gene1";gene_info="gene1";transcript_id="transcript2"
chr12   HAVANA  exon    100 600 .   +   .   ID="ccc2";Parent="bbb2";common_tag="gene1"
chr12   HAVANA  CDS 100 600 .   +   0   ID="ddd2";Parent="bbb2";common_tag="gene1"
chr12   HAVANA  gene    1000    5000    .   +   .   ID=nbisL1-transcript-3;common_tag="gene2";gene_info="gene2";transcript_id="transcript3"
chr12   HAVANA  transcript  1000    5000    .   +   .   ID="yyy";Parent=nbisL1-transcript-3;common_tag="gene2";gene_info="gene2";transcript_id="transcript3"
chr12   HAVANA  exon    1000    5000    .   +   .   ID="zzz";Parent="yyy";common_tag="gene2"
chr12   HAVANA  CDS 1000    5000    .   +   0   ID="www";Parent="yyy";common_tag="gene2"
</code></pre> <ul> <li>! A way to fix that is to use a common attribute (i.e. locus tag). AGAT uses <code>locus_tag</code> and <code>gene_id</code> by default. If you are lucky those attributes already exist. Here they are absent, you can use either <code>common_tag</code>, <code>transcript_id</code>, or <code>gene_info</code>. Let's investigate each case:</li> </ul> <p><code>agat config --expose --locus_tag common_tag # Modify the locus_tag parameter via the AGAT configuration file agat_config.yaml</code> <code>agat_convert_sp_gxf2gxf.pl --gff testA.gff</code></p> <p>This will work well even if transcript isoforms exist. This will use the parsing approach 2 (only using common attribute).</p> <pre><code>chr12   HAVANA  gene    100 600 .   +   .   ID=nbisL1-transcript-1;common_tag="gene1";gene_info="gene1";transcript_id="transcript1"
chr12   HAVANA  transcript  100 500 .   +   .   ID="bbb";Parent=nbisL1-transcript-1;common_tag="gene1";gene_info="gene1";transcript_id="transcript1"
chr12   HAVANA  exon    100 500 .   +   .   ID="ccc";Parent="bbb";common_tag="gene1"
chr12   HAVANA  CDS 100 500 .   +   0   ID="ddd";Parent="bbb";common_tag="gene1"
chr12   HAVANA  transcript  100 600 .   +   .   ID="bbb2";Parent=nbisL1-transcript-1;common_tag="gene1";gene_info="gene1";transcript_id="transcript2"
chr12   HAVANA  exon    100 600 .   +   .   ID="ccc2";Parent="bbb2";common_tag="gene1"
chr12   HAVANA  CDS 100 600 .   +   0   ID="ddd2";Parent="bbb2";common_tag="gene1"
chr12   HAVANA  gene    1000    5000    .   +   .   ID=nbisL1-transcript-2;common_tag="gene2";gene_info="gene2";transcript_id="transcript3"
chr12   HAVANA  transcript  1000    5000    .   +   .   ID="yyy";Parent=nbisL1-transcript-2;common_tag="gene2";gene_info="gene2";transcript_id="transcript3"
chr12   HAVANA  exon    1000    5000    .   +   .   ID="zzz";Parent="yyy";common_tag="gene2"
chr12   HAVANA  CDS 1000    5000    .   +   0   ID="www";Parent="yyy";common_tag="gene2"
</code></pre> <p><code>agat config --expose --locus_tag gene_info # Modify the locus_tag parameter via the AGAT configuration file agat_config.yaml</code> <code>agat_convert_sp_gxf2gxf.pl --gff testA.gff</code></p> <p>This will work well even if transcript isoforms exist. This will use the parsing approach 2 (common attribute gene_info) for transcript features and approach 3 (sequential) for subfeatures, which do not have the transcript_id attribute.</p> <pre><code>chr12   HAVANA  gene    100 600 .   +   .   ID="gene1";common_tag="gene1";gene_info="gene1";transcript_id="transcript1"
chr12   HAVANA  transcript  100 500 .   +   .   ID="bbb";Parent="gene1";common_tag="gene1";gene_info="gene1";transcript_id="transcript1"
chr12   HAVANA  exon    100 500 .   +   .   ID="ccc";Parent="bbb";common_tag="gene1"
chr12   HAVANA  CDS 100 500 .   +   0   ID="ddd";Parent="bbb";common_tag="gene1"
chr12   HAVANA  transcript  100 600 .   +   .   ID="bbb2";Parent="gene1";common_tag="gene1";gene_info="gene1";transcript_id="transcript2"
chr12   HAVANA  exon    100 600 .   +   .   ID="ccc2";Parent="bbb2";common_tag="gene1"
chr12   HAVANA  CDS 100 600 .   +   0   ID="ddd2";Parent="bbb2";common_tag="gene1"
chr12   HAVANA  gene    1000    5000    .   +   .   ID="gene2";common_tag="gene2";gene_info="gene2";transcript_id="transcript3"
chr12   HAVANA  transcript  1000    5000    .   +   .   ID="yyy";Parent="gene2";common_tag="gene2";gene_info="gene2";transcript_id="transcript3"
chr12   HAVANA  exon    1000    5000    .   +   .   ID="zzz";Parent="yyy";common_tag="gene2"
chr12   HAVANA  CDS 1000    5000    .   +   0   ID="www";Parent="yyy";common_tag="gene2"
</code></pre> <p><code>agat config --expose --locus_tag transcript_id # Modify the locus_tag parameter via the AGAT configuration file agat_config.yaml</code> <code>agat_convert_sp_gxf2gxf.pl --gff testA.gff</code> </p> <p>/!\ In our case, using <code>transcript_id</code> is not a good choice. Indeed each transcript will have its own gene feature, so isoform will not be linked to the same gene feature as expected. This will use the parsing approach 2 (common attribute transcript_id) for transcript features and approach 3 (sequential) for subfeatures that do not have the transcript_id attribute.</p> <pre><code>chr12    HAVANA  gene    100 500 .   +   .   ID="transcript1";common_tag="gene1";gene_info="gene1";transcript_id="transcript1"
chr12   HAVANA  transcript  100 500 .   +   .   ID="bbb";Parent="transcript1";common_tag="gene1";gene_info="gene1";transcript_id="transcript1"
chr12   HAVANA  exon    100 500 .   +   .   ID="ccc";Parent="bbb";common_tag="gene1"
chr12   HAVANA  CDS 100 500 .   +   0   ID="ddd";Parent="bbb";common_tag="gene1"
chr12   HAVANA  gene    100 600 .   +   .   ID="transcript2";common_tag="gene1";gene_info="gene1";transcript_id="transcript2"
chr12   HAVANA  transcript  100 600 .   +   .   ID="bbb2";Parent="transcript2";common_tag="gene1";gene_info="gene1";transcript_id="transcript2"
chr12   HAVANA  exon    100 600 .   +   .   ID="ccc2";Parent="bbb2";common_tag="gene1"
chr12   HAVANA  CDS 100 600 .   +   0   ID="ddd2";Parent="bbb2";common_tag="gene1"
chr12   HAVANA  gene    1000    5000    .   +   .   ID="transcript3";common_tag="gene2";gene_info="gene2";transcript_id="transcript3"
chr12   HAVANA  transcript  1000    5000    .   +   .   ID="yyy";Parent="transcript3";common_tag="gene2";gene_info="gene2";transcript_id="transcript3"
chr12   HAVANA  exon    1000    5000    .   +   .   ID="zzz";Parent="yyy";common_tag="gene2"
chr12   HAVANA  CDS 1000    5000    .   +   0   ID="www";Parent="yyy";common_tag="gene2"
</code></pre> <h4 id=b-level1-and-level2-feature-types-missing-only-level3-features>B. Level1 and Level2 feature types missing (Only Level3 features!)<a class=headerlink href=#b-level1-and-level2-feature-types-missing-only-level3-features title="Permanent link">#</a></h4> <p>In such case the sequential approach cannot be used (Indeed no level1 (e.g. gene) and no lelve2 (e.g. mrna) feature is present in the file). So the presence of parent/ID transcript_id/gene_id relationships and/or a proper common attribute is crucial.</p> <h5 id=1-case-with-parentid-transcript_idgene_id-relationships>1. Case with Parent/ID transcript_id/gene_id relationships.<a class=headerlink href=#1-case-with-parentid-transcript_idgene_id-relationships title="Permanent link">#</a></h5> <p>If you have isoforms (for Eukaryote organism) in your files and the <code>common attribute</code> used is not set properly you can end up with isoforms having independent parent gene features. See below for more details.</p> <p><strong>1.1</strong></p> <p>Input (testB.gff): </p> <pre><code>    chr12   HAVANA  exon    100 500 .   +   .   ID=exon1;Parent=transcript1;locus_id="gene1"
    chr12   HAVANA  CDS 100 500 .   +   0   ID=cds-1;Parent=transcript1;locus_id="gene1"
    chr12   HAVANA  exon    100 600 .   +   .   ID=exon2;Parent=transcript2;locus_id="gene1"
    chr12   HAVANA  CDS 100 600 .   +   0   ID=cds-2;Parent=transcript2;locus_id="gene1"
    chr12   HAVANA  exon    700 900 .   +   .   ID=exonb;Parent=transcriptb;locus_id="gene2"
    chr12   HAVANA  CDS 700 900 .   +   0   ID=cds-b;Parent=transcriptb;locus_id="gene2"
</code></pre> <ul> <li>/!\ Be careful in Eukaryote annotation containing isoforms. Indeed there is no Leve2 feature (e.g. mRNA) to indicate to which parental gene to link isoforms to. By default (see below) the result will be wrong because transcript1 and transcript2 should be attached to the same gene: </li> </ul> <p><code>agat_convert_sp_gxf2gxf.pl --gff testB.gff</code></p> <pre><code>    chr12   HAVANA  gene    100 500 .   +   .   ID=nbis-gene-1;locus_id="gene1"
    chr12   HAVANA  mRNA    100 500 .   +   .   ID=transcript1;Parent=nbis-gene-1;locus_id="gene1"
    chr12   HAVANA  exon    100 500 .   +   .   ID=exon1;Parent=transcript1;locus_id="gene1"
    chr12   HAVANA  CDS 100 500 .   +   0   ID=cds-1;Parent=transcript1;locus_id="gene1"
    chr12   HAVANA  gene    100 600 .   +   .   ID=nbis-gene-2;locus_id="gene1"
    chr12   HAVANA  mRNA    100 600 .   +   .   ID=transcript2;Parent=nbis-gene-2;locus_id="gene1"
    chr12   HAVANA  exon    100 600 .   +   .   ID=exon2;Parent=transcript2;locus_id="gene1"
    chr12   HAVANA  CDS 100 600 .   +   0   ID=cds-2;Parent=transcript2;locus_id="gene1"
    chr12   HAVANA  gene    700 900 .   +   .   ID=nbis-gene-3;locus_id="gene2"
    chr12   HAVANA  mRNA    700 900 .   +   .   ID=transcriptb;Parent=nbis-gene-3;locus_id="gene2"
    chr12   HAVANA  exon    700 900 .   +   .   ID=exonb;Parent=transcriptb;locus_id="gene2"
    chr12   HAVANA  CDS 700 900 .   +   0   ID=cds-b;Parent=transcriptb;locus_id="gene2"
</code></pre> <ul> <li>! A way to fix that is to use a <code>common attribute</code> to group the feature properly: AGAT uses <code>locus_tag</code> and <code>gene_id</code> by default. If you are lucky those attributes already exist. Here they are absent, you can use <code>locus_id</code> instead.</li> </ul> <p><code>agat config --expose --locus_tag locus_id # Modify the locus_tag parameter via the AGAT configuration file agat_config.yaml</code> <code>agat_convert_sp_gxf2gxf.pl --gff testB.gff</code></p> <pre><code>chr12   HAVANA  gene    100 600 .   +   .   ID="gene1";locus_id="gene1"
chr12   HAVANA  mRNA    100 500 .   +   .   ID=transcript1;Parent="gene1";locus_id="gene1"
chr12   HAVANA  exon    100 500 .   +   .   ID=exon1;Parent=transcript1;locus_id="gene1"
chr12   HAVANA  CDS 100 500 .   +   0   ID=cds-1;Parent=transcript1;locus_id="gene1"
chr12   HAVANA  mRNA    100 600 .   +   .   ID=transcript2;Parent="gene1";locus_id="gene1"
chr12   HAVANA  exon    100 600 .   +   .   ID=exon2;Parent=transcript2;locus_id="gene1"
chr12   HAVANA  CDS 100 600 .   +   0   ID=cds-2;Parent=transcript2;locus_id="gene1"
chr12   HAVANA  gene    700 900 .   +   .   ID="gene2";locus_id="gene2"
chr12   HAVANA  mRNA    700 900 .   +   .   ID=transcriptb;Parent="gene2";locus_id="gene2"
chr12   HAVANA  exon    700 900 .   +   .   ID=exonb;Parent=transcriptb;locus_id="gene2"
chr12   HAVANA  CDS 700 900 .   +   0   ID=cds-b;Parent=transcriptb;locus_id="gene2"
</code></pre> <p><strong>1.2</strong></p> <p>Here we have only level3 features, Parent/ID transcript_id/gene_id relationships present, default <code>common attributes</code> ( <code>locus_tag</code> or <code>gene_id</code>) is set for some features.</p> <p>Input testF.gff:</p> <pre><code>    chr12   HAVANA  exon    100 500 .   +   .   ID=exon1;Parent=transcript1;locus_tag="gene1"
    chr12   HAVANA  CDS 100 500 .   +   0   ID=cds-1;Parent=transcript1;locus_tag="gene1"
    chr12   HAVANA  exon    100 600 .   +   .   ID=exon2;Parent=transcript2;locus_tag="gene1"
    chr12   HAVANA  CDS 100 600 .   +   0   ID=cds-2;Parent=transcript2;locus_tag="gene1"
    chr12   HAVANA  exon    700 900 .   +   .   ID=exonb;Parent=transcriptb;locus_tag="gene2"
    chr12   HAVANA  CDS 700 900 .   +   0   ID=cds-b;Parent=transcriptb;locus_tag="gene2"
    chr12   HAVANA  exon    1000    1110    .   +   .   ID=exon4;Parent=transcript4
    chr12   HAVANA  CDS 1000    1110    .   +   0   ID=cds4;Parent=transcript4
</code></pre> <p><code>agat_convert_sp_gxf2gxf.pl --gff testF.gff</code></p> <pre><code>    chr12   HAVANA  gene    100 600 .   +   .   ID="gene1";locus_tag="gene1"
    chr12   HAVANA  mRNA    100 500 .   +   .   ID=transcript1;Parent="gene1";locus_tag="gene1"
    chr12   HAVANA  exon    100 500 .   +   .   ID=exon1;Parent=transcript1;locus_tag="gene1"
    chr12   HAVANA  CDS 100 500 .   +   0   ID=cds-1;Parent=transcript1;locus_tag="gene1"
    chr12   HAVANA  mRNA    100 600 .   +   .   ID=transcript2;Parent="gene1";locus_tag="gene1"
    chr12   HAVANA  exon    100 600 .   +   .   ID=exon2;Parent=transcript2;locus_tag="gene1"
    chr12   HAVANA  CDS 100 600 .   +   0   ID=cds-2;Parent=transcript2;locus_tag="gene1"
    chr12   HAVANA  gene    700 900 .   +   .   ID="gene2";locus_tag="gene2"
    chr12   HAVANA  mRNA    700 900 .   +   .   ID=transcriptb;Parent="gene2";locus_tag="gene2"
    chr12   HAVANA  exon    700 900 .   +   .   ID=exonb;Parent=transcriptb;locus_tag="gene2"
    chr12   HAVANA  CDS 700 900 .   +   0   ID=cds-b;Parent=transcriptb;locus_tag="gene2"
    chr12   HAVANA  gene    1000    1110    .   +   .   ID=nbis-gene-1
    chr12   HAVANA  mRNA    1000    1110    .   +   .   ID=transcript4;Parent=nbis-gene-1
    chr12   HAVANA  exon    1000    1110    .   +   .   ID=exon4;Parent=transcript4
    chr12   HAVANA  CDS 1000    1110    .   +   0   ID=cds4;Parent=transcript4
</code></pre> <p>The <code>common attributes</code> is used to attach isoforms to a common gene feature. As transcript4 has no common attribute, it will have its own parent features.</p> <h5 id=2-case-without-parentid-transcript_idgene_id-relationships-only-common-attribute-approach-to-parse-the-file-can-be-used>2. Case without Parent/ID transcript_id/gene_id relationships. Only <code>common attribute</code> approach to parse the file can be used.<a class=headerlink href=#2-case-without-parentid-transcript_idgene_id-relationships-only-common-attribute-approach-to-parse-the-file-can-be-used title="Permanent link">#</a></h5> <p><strong>2.1</strong></p> <p>Here we have only level3 features, no Parent/ID transcript_id/gene_id relationships, but a default <code>common attributes</code> ( <code>locus_tag</code> or <code>gene_id</code>) is present.</p> <p>Input testE.gff:</p> <pre><code>    chr12   HAVANA  exon    100 300 .   +   .   ID=exon1;locus_tag="gene1"
    chr12   HAVANA  CDS 100 300 .   +   0   ID=cds-1;locus_tag="gene1"
    chr12   HAVANA  exon    500 600 .   +   .   ID=exon2;locus_tag="gene1"
    chr12   HAVANA  CDS 500 600 .   +   0   ID=cds-2;locus_tag="gene1"
    chr12   HAVANA  exon    700 900 .   +   .   ID=exonb;locus_tag="gene2"
    chr12   HAVANA  CDS 700 900 .   +   0   ID=cds-b;locus_tag="gene2"
</code></pre> <p><code>agat_convert_sp_gxf2gxf.pl --gff testE.gff</code></p> <pre><code>    chr12   HAVANA  gene    100 600 .   +   .   ID=nbis-gene-1;locus_tag="gene1"
    chr12   HAVANA  mRNA    100 600 .   +   .   ID=nbisL2-exon-1;Parent=nbis-gene-1;locus_tag="gene1"
    chr12   HAVANA  exon    100 300 .   +   .   ID=exon1;Parent=nbisL2-exon-1;locus_tag="gene1"
    chr12   HAVANA  exon    500 600 .   +   .   ID=exon2;Parent=nbisL2-exon-1;locus_tag="gene1"
    chr12   HAVANA  CDS 100 300 .   +   0   ID=cds-1;Parent=nbisL2-exon-1;locus_tag="gene1"
    chr12   HAVANA  CDS 500 600 .   +   0   ID=cds-2;Parent=nbisL2-exon-1;locus_tag="gene1"
    chr12   HAVANA  gene    700 900 .   +   .   ID=nbis-gene-2;locus_tag="gene2"
    chr12   HAVANA  mRNA    700 900 .   +   .   ID=nbisL2-exon-2;Parent=nbis-gene-2;locus_tag="gene2"
    chr12   HAVANA  exon    700 900 .   +   .   ID=exonb;Parent=nbisL2-exon-2;locus_tag="gene2"
    chr12   HAVANA  CDS 700 900 .   +   0   ID=cds-b;Parent=nbisL2-exon-2;locus_tag="gene2"
</code></pre> <p>/!\ In Eukaryote annotation containing isoforms it will not work properly. Indeed, it will result of isoforms merged in chimeric transcripts (It will be really unlucky to end up in such situation, because even a human cannot resolve such type of situation. There is no information about isoforms structure...). In Eukaryote cases (even for multi-exon CDS) with absence of isoforms, it will work correctly.</p> <p>2.2</p> <p>Here the worse case that can append: only level3 features, no Parent/ID transcript_id/gene_id relationships, and the default <code>common attributes</code> ( <code>locus_tag</code> and <code>gene_id</code>) are absent. Sequential approach will be used by AGAT but as there are only level3 features, all will be linked to only one parent. See below for more details.</p> <p>Input testC.gff: </p> <pre><code>chr12   HAVANA  exon    100 500 .   +   .   ID=exon1;locus_id="gene1"
chr12   HAVANA  CDS 100 500 .   +   0   ID=cds-1;locus_id="gene1"
chr12   HAVANA  exon    510 600 .   +   .   ID=exon2;locus_id="gene1"
chr12   HAVANA  CDS 510 600 .   +   0   ID=cds-2;locus_id="gene1"
chr12   HAVANA  exon    700 900 .   +   .   ID=exonb;locus_id="gene2"
chr12   HAVANA  CDS 700 900 .   +   0   ID=cds-b;locus_id="gene2"
</code></pre> <p><code>agat_convert_sp_gxf2gxf.pl --gff testC.gff</code> </p> <pre><code>chr12   HAVANA  gene    100 900 .   +   .   ID=nbis-gene-1;locus_id="gene1"
chr12   HAVANA  mRNA    100 900 .   +   .   ID=nbisL2-exon-1;Parent=nbis-gene-1;locus_id="gene1"
chr12   HAVANA  exon    100 600 .   +   .   ID=exon1;Parent=nbisL2-exon-1;locus_id="gene1"
chr12   HAVANA  exon    700 900 .   +   .   ID=exonb;Parent=nbisL2-exon-1;plocus_id="gene2"
chr12   HAVANA  CDS 100 500 .   +   0   ID=cds-1;Parent=nbisL2-exon-1;locus_id="gene1"
chr12   HAVANA  CDS 100 600 .   +   0   ID=cds-2;Parent=nbisL2-exon-1;locus_id="gene1"
chr12   HAVANA  CDS 700 900 .   +   0   ID=cds-b;Parent=nbisL2-exon-1;locus_id="gene2"
</code></pre> <p>/!\ All features are collected under a single gene and mRNA feature, which is wrong.</p> <p>As the default <code>common attribute</code> are absent (gene_id or locus_tag), you have to inform AGAT what attribute to use to group features together properly, here <code>locus_id</code> is a good choice: </p> <p><code>agat config --expose --locus_tag locus_id # Modify the locus_tag parameter via the AGAT configuration file agat_config.yaml</code> <code>agat_convert_sp_gxf2gxf.pl --gff testC.gff</code> </p> <pre><code>chr12   HAVANA  gene    100 600 .   +   .   ID=nbis-gene-1;locus_id="gene1"
chr12   HAVANA  mRNA    100 600 .   +   .   ID=nbisL2-exon-1;Parent=nbis-gene-1;locus_id="gene1"
chr12   HAVANA  exon    100 600 .   +   .   ID=exon1;Parent=nbisL2-exon-1;locus_id="gene1"
chr12   HAVANA  CDS 100 500 .   +   0   ID=cds-1;Parent=nbisL2-exon-1;locus_id="gene1"
chr12   HAVANA  CDS 100 600 .   +   0   ID=cds-2;Parent=nbisL2-exon-1;locus_id="gene1"
chr12   HAVANA  gene    700 900 .   +   .   ID=nbis-gene-2;locus_id="gene2"
chr12   HAVANA  mRNA    700 900 .   +   .   ID=nbisL2-exon-2;Parent=nbis-gene-2;locus_id="gene2"
chr12   HAVANA  exon    700 900 .   +   .   ID=exonb;Parent=nbisL2-exon-2;locus_id="gene2"
chr12   HAVANA  CDS 700 900 .   +   0   ID=cds-b;Parent=nbisL2-exon-2;locus_id="gene2"
</code></pre> <p>/!\ In Eukaryote annotation containing isoforms it will not work properly. Indeed, it will result of isoforms merged in chimeric transcripts (It will be really unlucky to end up in such situation, because even a human cannot resolve such type of situation. There is no information about isoforms structure...). In Eukaryote cases (even for multi-exon CDS) with absence of isoforms, it will work correctly.</p> <h5 id=3-in-the-extreme-case-where-you-have-only-one-type-of-feature-you-may-decide-to-use-the-id-as-common-attribute>3. In the extreme case where you have only one type of feature, you may decide to use the ID as common attribute.<a class=headerlink href=#3-in-the-extreme-case-where-you-have-only-one-type-of-feature-you-may-decide-to-use-the-id-as-common-attribute title="Permanent link">#</a></h5> <p>This is the same problem as seen previously. Here the worse case that can append: only level3 features, no Parent/ID transcript_id/gene_id relationships, and the default <code>common attributes</code> ( <code>locus_tag</code> and <code>gene_id</code>) are absent. Sequential approach will be used by AGAT but as there are only level3 features, all will be linked to only one parent. See below for more details.</p> <p>Input (testD.gff):</p> <pre><code>    chr10   Liftoff CDS 100 300 .   +   0   ID=cds1
    chr10   Liftoff CDS 600 900 .   +   0   ID=cds2
    chr10   Liftoff CDS 400 490 .   -   0   ID=cds3
</code></pre> <p><code>agat_convert_sp_gxf2gxf.pl --gff testD.gff</code> </p> <pre><code>    chr10   Liftoff gene    100 900 .   +   .   ID=nbis-gene-1
    chr10   Liftoff mRNA    100 900 .   +   .   ID=nbisL2-cds-1;Parent=nbis-gene-1
    chr10   Liftoff exon    100 300 .   +   .   ID=nbis-exon-1;Parent=nbisL2-cds-1
    chr10   Liftoff exon    400 490 .   +   .   ID=nbis-exon-2;Parent=nbisL2-cds-1
    chr10   Liftoff exon    600 900 .   +   .   ID=nbis-exon-3;Parent=nbisL2-cds-1
    chr10   Liftoff CDS 100 300 .   +   0   ID=cds1;Parent=nbisL2-cds-1
    chr10   Liftoff CDS 400 490 .   -   0   ID=cds3;Parent=nbisL2-cds-1
    chr10   Liftoff CDS 600 900 .   +   0   ID=cds2;Parent=nbisL2-cds-1
</code></pre> <p>/!\ All features are collected under a single gene and mRNA feature, which is wrong.</p> <p><code>agat config --expose --locus_tag ID # Modify the locus_tag parameter via the AGAT configuration file agat_config.yaml</code> <code>agat_convert_sp_gxf2gxf.pl --gff testD.gff</code></p> <pre><code>    chr10   Liftoff gene    100 300 .   +   0   ID=nbis-gene-1
    chr10   Liftoff mRNA    100 300 .   +   0   ID=nbisL2-cds-1;Parent=nbis-gene-1
    chr10   Liftoff exon    100 300 .   +   .   ID=nbis-exon-1;Parent=nbisL2-cds-1
    chr10   Liftoff CDS 100 300 .   +   0   ID=cds1;Parent=nbisL2-cds-1
    chr10   Liftoff gene    400 490 .   -   0   ID=nbis-gene-3
    chr10   Liftoff mRNA    400 490 .   -   0   ID=nbisL2-cds-3;Parent=nbis-gene-3
    chr10   Liftoff exon    400 490 .   -   .   ID=nbis-exon-3;Parent=nbisL2-cds-3
    chr10   Liftoff CDS 400 490 .   -   0   ID=cds3;Parent=nbisL2-cds-3
    chr10   Liftoff gene    600 900 .   +   0   ID=nbis-gene-2
    chr10   Liftoff mRNA    600 900 .   +   0   ID=nbisL2-cds-2;Parent=nbis-gene-2
    chr10   Liftoff exon    600 900 .   +   .   ID=nbis-exon-2;Parent=nbisL2-cds-2
    chr10   Liftoff CDS 600 900 .   +   0   ID=cds2;Parent=nbisL2-cds-2
</code></pre> <p>This case is fine for Prokaryote annotation.<br> /!\ For Eukaryote it might work is some conditions:<br> A) The annotation should not contain isoforms (Indeed, there is no existing information to decipher to which isoform a CDS will be part of. If isoforms are present, each one will be linked to its own gene feature).<br> B) If there are multi-exon CDS, CDS parts must share the same ID (Indeed multi-exon CDS can share or not the same ID. Both way are allowed by the GFF format. If the CDS parts share the same ID, the CDS parts will be collected properly. If the CDS parts do not share the same ID, AGAT will slice it and create a gene/mRNA feature by CDS part!).</p> <h5 id=4-case-where-you-have-only-one-type-of-feature-and-some-feature-have-parent-attributes-and-some-other-have-common-attributes>4. Case where you have only one type of feature, and some feature have Parent attributes and some other have common attributes.<a class=headerlink href=#4-case-where-you-have-only-one-type-of-feature-and-some-feature-have-parent-attributes-and-some-other-have-common-attributes title="Permanent link">#</a></h5> <p>Input (testG.gff):</p> <pre><code>    chr12   HAVANA  exon    100 500 .   +   .   ID=exon1;Parent=transcript1
    chr12   HAVANA  CDS 100 500 .   +   0   ID=cds-1;Parent=transcript1
    chr12   HAVANA  exon    100 600 .   +   .   ID=exon2;Parent=transcript2
    chr12   HAVANA  CDS 100 600 .   +   0   ID=cds-2;Parent=transcript2
    chr12   HAVANA  exon    700 900 .   +   .   ID=exonb;locus_tag="gene1"
    chr12   HAVANA  CDS 700 900 .   +   0   ID=cds-b;locus_tag="gene1"
    chr12   HAVANA  exon    1000    1110    .   +   .   ID=exon4;locus_tag="gene2"
    chr12   HAVANA  CDS 1000    1110    .   +   0   ID=cds4;locus_tag="gene2"
</code></pre> <p><code>agat_convert_sp_gxf2gxf.pl --gff testG.gff</code> </p> <pre><code>    chr12   HAVANA  gene    100 500 .   +   .   ID=nbis-gene-3
    chr12   HAVANA  mRNA    100 500 .   +   .   ID=transcript1;Parent=nbis-gene-3
    chr12   HAVANA  exon    100 500 .   +   .   ID=exon1;Parent=transcript1
    chr12   HAVANA  CDS 100 500 .   +   0   ID=cds-1;Parent=transcript1
    chr12   HAVANA  gene    100 600 .   +   .   ID=nbis-gene-4
    chr12   HAVANA  mRNA    100 600 .   +   .   ID=transcript2;Parent=nbis-gene-4
    chr12   HAVANA  exon    100 600 .   +   .   ID=exon2;Parent=transcript2
    chr12   HAVANA  CDS 100 600 .   +   0   ID=cds-2;Parent=transcript2
    chr12   HAVANA  gene    700 900 .   +   .   ID=nbis-gene-1;locus_tag="gene1"
    chr12   HAVANA  mRNA    700 900 .   +   .   ID=nbisL2-exon-1;Parent=nbis-gene-1;locus_tag="gene1"
    chr12   HAVANA  exon    700 900 .   +   .   ID=exonb;Parent=nbisL2-exon-1;locus_tag="gene1"
    chr12   HAVANA  CDS 700 900 .   +   0   ID=cds-b;Parent=nbisL2-exon-1;locus_tag="gene1"
    chr12   HAVANA  gene    1000    1110    .   +   .   ID=nbis-gene-2;locus_tag="gene2"
    chr12   HAVANA  mRNA    1000    1110    .   +   .   ID=nbisL2-exon-2;Parent=nbis-gene-2;locus_tag="gene2"
    chr12   HAVANA  exon    1000    1110    .   +   .   ID=exon4;Parent=nbisL2-exon-2;locus_tag="gene2"
    chr12   HAVANA  CDS 1000    1110    .   +   0   ID=cds4;Parent=nbisL2-exon-2;locus_tag="gene2"
</code></pre> <p>/!\ For Eukaryote annotation with isoforms, features would need to have the Parent attribute along with a common attribute to help AGAT to properly reconstruct the parental features (a single gene feature for isoforms).</p> </article> </div> <script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script> </div> <button type=button class="md-top md-icon" data-md-component=top hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg> Back to top </button> </main> <footer class=md-footer> <nav class="md-footer__inner md-grid" aria-label=Footer> <a href=../agat_for_you/ class="md-footer__link md-footer__link--prev" aria-label="Previous: What can AGAT do for you?"> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg> </div> <div class=md-footer__title> <span class=md-footer__direction> Previous </span> <div class=md-ellipsis> What can AGAT do for you? </div> </div> </a> <a href=../how_to_cite/ class="md-footer__link md-footer__link--next" aria-label="Next: How to cite?"> <div class=md-footer__title> <span class=md-footer__direction> Next </span> <div class=md-ellipsis> How to cite? </div> </div> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11z"/></svg> </div> </a> </nav> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"base": "..", "features": ["announce.dismiss", "content.action.edit", "content.action.view", "content.code.annotate", "content.code.copy", "content.tooltips", "header.autohide", "navigation.footer", "navigation.indexes", "navigation.tabs", "navigation.top", "navigation.tracking", "search.highlight", "search.share", "search.suggest", "toc.follow", "toc.integrate"], "search": "../assets/javascripts/workers/search.d50fe291.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script> <script src=../assets/javascripts/bundle.50899def.min.js></script> </body> </html>